src/backend/distributed/commands/dependencies.c

```diff
@@ -0,0 +1,196 @@
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "catalog/dependency.h"
+#include "catalog/objectaddress.h"
+
+#include "distributed/dist_catalog/dependency.h"
+#include "distributed/connection_management.h"
+#include "distributed/worker_manager.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/commands.h"
+#include "distributed/remote_commands.h"
+#include "distributed/dist_catalog/distobject.h"
+
+static List * GetDependencyCreateDDLCommands(const ObjectAddress *dependency);
+
+
+/*
+ * EnsureDependenciesExists finds all the dependencies that we support and makes sure
+ * these are available on all workers. If not available they will be created on the
+ * workers via a separate session that will be committed directly so that the objects are
+ * visible to potentially multiple sessions creating the shards.
+ *
+ * Note; only the actual objects are created via a separate session, the local records to
+ * pg_dist_object are created in this session. As a side effect the objects could be
+ * created on the workers without a catalog entry on the coordinator. Updates to the
+ * objects on the coordinator are not propagated to the workers until the record is
+ * visible on the coordinator.
+ *
+ * This is solved by creating the dependencies in an idempotent manner, either via
+ * postgres native CREATE IF NOT EXISTS, or citus helper functions.
+ */
+void
+EnsureDependenciesExistsOnAllNodes(const ObjectAddress *target)
+{
+	const uint64 connectionFlag = FORCE_NEW_CONNECTION;
+
+	/* local variables to work with dependencies */
+	List *dependencies = NIL;
+	ListCell *dependencyCell = NULL;
+
+	/* local variables to collect ddl commands */
+	List *ddlCommands = NULL;
+
+	/* local variables to work with worker nodes */
+	List *workerNodeList = NULL;
+	ListCell *workerNodeCell = NULL;
+	List *connections = NULL;
+	ListCell *connectionCell = NULL;
+
+	/*
+	 * collect all dependencies in creation order and get their ddl commands
+	 */
+	dependencies = GetDependenciesForObject(target);
+	foreach(dependencyCell, dependencies)
+	{
+		ObjectAddress *dependency = (ObjectAddress *) lfirst(dependencyCell);
+		ddlCommands = list_concat(ddlCommands,
+								  GetDependencyCreateDDLCommands(dependency));
+	}
+	if (list_length(ddlCommands) <= 0)
+	{
+		/* no ddl commands to be executed */
+		return;
+	}
+
+	/*
+	 * collect and connect to all applicable nodes
+	 */
+	workerNodeList = ActivePrimaryNodeList();
+	if (list_length(workerNodeList) <= 0)
+	{
+		/* no nodes to execute on */
+		return;
+	}
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		MultiConnection *connection = NULL;
+
+		char *nodeName = workerNode->workerName;
+		uint32 nodePort = workerNode->workerPort;
+
+		connection = GetNodeUserDatabaseConnection(connectionFlag, nodeName,
```

Comment:
Minor: Could be a nice opportunity to use StartNodeUserDatabaseConnection+FinishConnectionListEstablishment. create_distributed_table can be relatively frequent.

We should perhaps also look for ways to do this in the executor.

Reply:
Just for my understanding, (where) does the executor fail right now? We seem to handle INVALID_SHARD_ID in placement_connection.c. We have other cases in which tasks don't have shard or placement IDs that go through the executor.

Reply:
Hmm, actually that's true. I've misread some parts of the code. Still, I think there are some things that we should be aware of such as:

We should be switching sequential mode after executing this task, otherwise `create_distributed_table()` would get into distributed deadlocks in the same tx. And, those would error out with if the table has data in it.
```
< ERROR:  cannot distribute "partitioning_test" in sequential mode because it is not empty
< HINT:  If you have manually set citus.multi_shard_modify_mode to 'sequential', try with 'parallel' option. If that is not the case, try distributing local tables when they are empty.
---
```

Well, COPY is actually able to handle such cases, we probably had forgotten some checks.

Here is a prototype for using the executor: https://github.com/citusdata/citus/commit/6d54e1eb84356e780b1dc4988abdcfdbe2edd1e3


Do you think is it acceptable to switch to sequential mode?

Reply:
This call site is different from calling it during the creation of the type.

These commands need to be executed on a new connection and commit before any shards get created. That is to make the objects visible to all connections used with `create_distributed_table`. Would this at all be possible with the executor?

Reply:
> These commands need to be executed on a new connection and commit before any shards get created.

If we use sequential mode (e.g., 1 connection per node), we should be able to handle this. But then we probably wouldn't want that because of performance considerations.

And, also we should error out if one (or more) parallel queries have already been executed in the same transaction block earlier -- if we're in a tx block. So, basically enforce similar restrictions with foreign keys to reference tables.

Overall, that looks undesirable unless @marcocitus has a better suggestion that I cannot see now.

Reply:
I've had a lot of discussions on this topic with Nils. Committing separately is the right approach for now.

Parallelism is one reason. We also need to ensure that the type had the right owner and at the same time make it visible to the current user to create shards. In addition, we also need superuser to be able to see the type to create the distributed table and write the metadata on the worker during create_distributed_table. There are some elaborate schemes we could use to circumvent this, but that would create a ton of extra work.

Reply:
I think the executor currently always uses the current user and the transactional behavior isn't configurable. Those should be easy to resolve, but I'm fine with punting on this given time constraints.

Reply:
> Minor: Could be a nice opportunity to use StartNodeUserDatabaseConnection+FinishConnectionListEstablishment. create_distributed_table can be relatively frequent.

I guess Marco's above note is still applicable.